A79-non-per-call-metrics-architecture.md

```diff
@@ -0,0 +1,405 @@
+example, third-party plugins to gRPC that register instruments might be
+dynamically loaded. Even if dynamic registration of instruments is supported,
+implementations can choose to add constraints such that newly registered
+instruments are only used by stats plugins that are registered thereafter.
+
+If multiple instrument descriptors are registered with the same name, it is
+considered an error and should result in a panic. This behavior might be relaxed
+in the future, for example, if the same instrument is registered again, we could
+return the previously registered instrument. This would be useful in cases where
+a library is dynamically loaded and unloaded multiple times during the lifetime
+of the process.
+
+![](A79_graphics/global-instruments-registry.png)
+
+##### Core
+
+C++, Python and other wrapped languages will use this API.
+
+```c++
+// A global registry of instruments(metrics). This API is designed to be used to
+// register instruments (Counter and Histogram) as part of program startup,
+// before the execution of the main function (during dynamic initialization
+// time). Using this API after the main function begins may result into missing
+// instruments. This API is thread-unsafe.
+class GlobalInstrumentsRegistry {
+ public:
+  enum class ValueType {
+    kUndefined,
+    kUInt64,
+    kDouble,
+  };
+  enum class InstrumentType {
+    kUndefined,
+    kCounter,
+    kHistogram,
+  };
+  struct GlobalInstrumentDescriptor {
+    ValueType value_type;
+    InstrumentType instrument_type;
+    uint32_t index;
+    bool enable_by_default;
+    absl::string_view name;
+    absl::string_view description;
+    absl::string_view unit;
+    std::vector<absl::string_view> label_keys;
+    std::vector<absl::string_view> optional_label_keys;
+  };
+  struct GlobalInstrumentHandle {
+    uint32_t index;
+  };
+  struct GlobalUInt64CounterHandle : public GlobalInstrumentHandle {};
+  struct GlobalDoubleHistogramHandle : public GlobalInstrumentHandle {};
+
+  // Creates instrument in the GlobalInstrumentsRegistry.
+  static GlobalUInt64CounterHandle RegisterUInt64Counter(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> optional_label_keys,
+      bool enable_by_default);
+  static GlobalDoubleHistogramHandle RegisterDoubleHistogram(
+      absl::string_view name, absl::string_view description,
+      absl::string_view unit, absl::Span<const absl::string_view> label_keys,
+      absl::Span<const absl::string_view> optional_label_keys,
+      bool enable_by_default);
+  static void ForEach(
+      absl::FunctionRef<void(const GlobalInstrumentDescriptor&)> f);
+};
+```
+
+##### Java
+
+(To be filled.)
+
+##### Go
+
+(To be filled.)
+
+#### Global Stats Plugin Registry
+
+A global stats plugin registry is added. This registry keeps track of all stats
```

Comment:
Didn't we say that Java and Go aren't going to have a global registry, or maybe they would have a global registry that can set only a single stats plugin?

Reply:
For Go no global registry, just the ability to set a global dial option (and eventually server option).